core/src/main/java/org/elasticsearch/index/fielddata/IndexFieldDataService.java

```diff
@@ -78,7 +79,7 @@ public IndexFieldDataService(IndexSettings indexSettings, IndicesFieldDataCache
         this.mapperService = mapperService;
     }
 
-    public synchronized void clear() {
+    public void clear() {
```

Comment:
This is not thread-safe. The problem is the underlying cache implementation that backs `IndexFieldCache` [says  @this](https://github.com/elastic/elasticsearch/blob/1e99195743a9a4d0cb637f4f7ceec92ab7867518/core/src/main/java/org/elasticsearch/common/cache/Cache.java#L553-L559):

```
    /**
     * An LRU sequencing of the keys in the cache that supports removal. This sequence is not protected from mutations
     * to the cache (except for {@link Iterator#remove()}. The result of iteration under any other mutation is
     * undefined.
     *
     * @return an LRU-ordered {@link Iterable} over the keys in the cache
     */
```

and this method is invoked behind the scenes to invalidate all the keys.

Reply:
added `synchronized` back